implementation/egads/src/main/java/net/opentsdb/query/anomaly/egads/olympicscoring/OlympicScoringNode.java

```diff
@@ -0,0 +1,815 @@
+    while (it.hasNext()) {
+      it.advance();
+      TimeSeries ts = it.value().predict(properties);
+      if (ts != null) {
+        computed.add(ts);
+      }
+    }
+    TimeStamp start = new SecondTimeStamp(prediction_start);
+    TimeStamp end = start.getCopy();
+    end.add(modelDuration() == ChronoUnit.HOURS ? Duration.ofHours(1) : Duration.ofDays(1));
+
+    if (cache != null && config.getMode() != ExecutionMode.CONFIG) {
+      // need's a clone as we may modify the list when we add thresholds, etc.
+      writeCache(new EgadsPredictionResult(this, 
+                                          data_source, 
+                                          start, 
+                                          end, 
+                                          Lists.newArrayList(computed), 
+                                          id_type));
+    }
+    prediction = new EgadsPredictionResult(this, 
+                                           data_source, 
+                                           start, 
+                                           end, 
+                                           computed, 
+                                           id_type);
+    countdown();
+  }
+  
+  class CacheCB implements Callback<Void, QueryResult> {
+
+    @Override
+    public Void call(final QueryResult result) throws Exception {
+      // TODO
+      // If result == null ? fire baseline, else store and use to match.
+      if (result != null) {
+        context.queryContext().logDebug("Prediction cache hit for query.");
+        prediction = result;
+        cache_hit = true;
+        countdown();
+      } else {
+        context.queryContext().logDebug("Prediction cache miss for query.");
+        fetchBaselineData();
+      }
+      
+      return null;
+    }
+    
+  }
+  
+  class CacheErrCB implements Callback<Void, Exception> {
+
+    @Override
+    public Void call(final Exception e) throws Exception {
+      LOG.warn("Cache exception", e);
+      fetchBaselineData();
+      return null;
+    }
+    
+  }
+  
+  class BaselineQuery implements QuerySink {
+    final int idx;
+    QueryContext sub_context;
+    
+    BaselineQuery(final int idx) {
+      this.idx = idx;
+    }
+    
+    @Override
+    public void onComplete() {
+      if (failed.get()) {
+        return;
+      }
+      
+      if (idx + 1 < baseline_queries.length) {
+        // fire next
+        baseline_queries[idx + 1].sub_context.initialize(null)
+          .addCallback(new SubQueryCB(baseline_queries[idx + 1].sub_context))
+          .addErrback(new ErrorCB());
+      } else {
+        runBaseline();
+      }
+    }
+
+    @Override
+    public void onNext(final QueryResult next) {
+      updateState(State.RUNNING, null);
+      // TODO filter, for now assume one result
+      
+      if (data_source == null) {
+        synchronized (OlympicScoringNode.this) {
+          if (data_source == null) {
+            data_source = next.dataSource();
```

Comment:
This DCLP enforces that one and only one thread ever writes to the `data_source` when it is `null`.